exonum-java-binding/core/src/test/java/com/exonum/binding/core/runtime/TransactionMethodExtractorTest.java

```diff
@@ -0,0 +1,185 @@
+    assertThat(e.getMessage()).contains(errorMessage);
+  }
+
+  @Test
+  void findMethodsValidServiceInterfaceImplementation() throws Exception {
+    Map<Integer, Method> transactions =
+        TransactionMethodExtractor.findTransactionMethods(
+            ValidServiceInterfaceImplementation.class);
+    assertThat(transactions).hasSize(2);
+    Method transactionMethod = ValidServiceInterfaceImplementation.class.getMethod(
+        "transactionMethod", byte[].class, TransactionContext.class);
+    Method transactionMethod2 = ValidServiceInterfaceImplementation.class.getMethod(
+        "transactionMethod2", byte[].class, TransactionContext.class);
+    List<Method> actualMethods = Arrays.asList(transactionMethod, transactionMethod2);
+    assertThat(actualMethods).containsExactlyInAnyOrderElementsOf(transactions.values());
+  }
+
+  class BasicService implements Service {
+
+    static final int TRANSACTION_ID = 1;
+
+    @Override
+    public List<HashCode> getStateHashes(Snapshot snapshot) {
+      return Collections.emptyList();
+    }
+
+    @Override
+    public void createPublicApiHandlers(Node node, Router router) {
+      // no-op
+    }
+  }
+
+  class ValidService extends BasicService {
+
+    @TransactionMethod(TRANSACTION_ID)
+    @SuppressWarnings("WeakerAccess") // Should be accessible
+    public void transactionMethod(byte[] arguments, TransactionContext context) {}
+  }
+
+  class DuplicateTransactionIdsService extends BasicService {
+
+    @TransactionMethod(TRANSACTION_ID)
+    public void transactionMethod(byte[] arguments, TransactionContext context) {}
+
+    @TransactionMethod(TRANSACTION_ID)
+    public void anotherTransactionMethod(byte[] arguments, TransactionContext context) {}
+  }
+
+  class MissingTransactionMethodArgumentsService extends BasicService {
+
+    @TransactionMethod(TRANSACTION_ID)
+    public void transactionMethod(byte[] arguments) {}
+  }
+
+  class InvalidTransactionMethodArgumentsService extends BasicService {
+
+    @TransactionMethod(TRANSACTION_ID)
+    public void transactionMethod(byte[] arguments, String invalidArgument) {}
+  }
+
+  class DuplicateTransactionMethodArgumentsService extends BasicService {
+
+    @TransactionMethod(TRANSACTION_ID)
+    public void transactionMethod(byte[] arguments, byte[] context) {}
+  }
+
+  interface ServiceInterface {
+    int TRANSACTION_ID = 1;
+
+    @TransactionMethod(TRANSACTION_ID)
+    void transactionMethod(byte[] arguments, TransactionContext context);
+  }
+
+  class ValidServiceInterfaceImplementation implements ServiceInterface {
+
+    static final int TRANSACTION_ID_2 = 2;
+
+    @TransactionMethod(TRANSACTION_ID)
+    public void transactionMethod(byte[] arguments, TransactionContext context) {}
+
+    @TransactionMethod(TRANSACTION_ID_2)
+    @SuppressWarnings("WeakerAccess") // Should be accessible
+    public void transactionMethod2(byte[] arguments, TransactionContext context) {}
```

Comment:
A bike-shed target (@slowli , @bullet-tooth , @MakarovS, @vitvakatu )! Shall we require the arguments to be in order:
1. (MessageT arguments, TransactionContext context)
2. (TransactionContext context, MessageT arguments)
3. Allow any (mind extra impl. complexity!)

Reply:
I think there is no need in arbitrary order. The either one works, but Rust runtime uses `context, arguments` one.

Reply:
I agree with allowing any order, there'd be a couple more lines of code in implementation, but less unnecessary restrictions for users.
Created ECR-3990

Reply:
I agree that it is not badly needed, and any could work: in some cases,  the `arguments` is the first argument used (e.g., to check preconditions); in some cases, the `context` (e.g., to instantiate the schema and check the preconditions _depending on the state_, or to just update it).

Reply:
No particular preference. From the unification perspective, `context, arguments` ordering makes sense, but AFAIK, this is not the Java way™. Since reflection is used to handle the processing, restricting the ordering of args is not necessary. (Further, it could make sense to allow to extract commonly used parts of the context into args, e.g., the service schema and the caller.)